src/engine/enginemaster.cpp

```diff
@@ -366,16 +406,80 @@ void EngineMaster::process(const int iBufferSize) {
     m_headphoneGain.setGain(chead_gain);
 
-    if (m_bRampingGain) {
-        ChannelMixer::mixChannelsRamping(
-                m_headphoneGain, &m_activeHeadphoneChannels,
-                &m_channelHeadphoneGainCache,
-                m_pHead, iBufferSize);
+    // Mixxx gets input signals delayed by the input latency,
+    // so to preserve the relative timing of the input with the master and headphones
+    // outputs, delay the master, booth, and headphones outputs by the input latency.
+    // This way, if the microphone user is on beat with the output they hear coming from
+    // the speakers/headphones, the input will remain on beat at the expense of
+    // adding some latency.
+    // If the input signal is being directly monitored (mixed with the master output
+    // in hardware), do not delay the master output. The recording/broadcasting signal
+    // will be delayed in this case to keep the recorded/broadcasted signal on beat
+    // and consistent with what the user hears out the speakers/headphones.
+    // If no microphone inputs are configured, avoid adding this extra latency.
+    TalkoverMixMode configuredTalkoverMixMode = static_cast<TalkoverMixMode>(
+        static_cast<int>(m_pTalkoverMixMode->get()));
+    // These latency values are in milliseconds as required by EngineDelay.
+    double inputLatencyCompensation;
+    double measuredRoundTripLatency = m_pRoundTripLatency->get();
+    // iBufferSize / iSampleRate gives double Mixxx's processing latency because
+    // there are 2 channels per buffer. FIXME when removing the assumption of stereo channels.
+    double processingLatency = (double)iBufferSize / iSampleRate / 2.0 * 1000.0;
+    if (measuredRoundTripLatency == 0.0) {
+        inputLatencyCompensation = processingLatency;
     } else {
-        ChannelMixer::mixChannels(
-                m_headphoneGain, &m_activeHeadphoneChannels,
-                &m_channelHeadphoneGainCache,
-                m_pHead, iBufferSize);
+        inputLatencyCompensation = (measuredRoundTripLatency - (processingLatency * 2)) / 2.0;
+    }
+    m_pInputLatencyCompensationDelay->setDelay(inputLatencyCompensation);
+    m_pInputLatencyCompensationHeadphonesDelay->setDelay(inputLatencyCompensation);
+
+    if (headphoneEnabled) {
+        if (m_bRampingGain) {
+            ChannelMixer::mixChannelsRamping(
+                    m_headphoneGain, &m_activeHeadphoneChannels,
+                    &m_channelHeadphoneGainCache,
+                    m_pHead, iBufferSize);
+        } else {
+            ChannelMixer::mixChannels(
+                    m_headphoneGain, &m_activeHeadphoneChannels,
+                    &m_channelHeadphoneGainCache,
+                    m_pHead, iBufferSize);
+        }
+
+        if (configuredTalkoverMixMode != TalkoverMixMode::DIRECT_MONITOR
+            && m_pNumMicsConfigured->get() > 0) {
+            m_pInputLatencyCompensationHeadphonesDelay->process(m_pHead, iBufferSize);
```

Comment:
Oh whoops, this is double processed. I had fixed this when merging this branch into my personal branch so didn't notice it before.